### PR TITLE
split avx from avx2

### PR DESCRIPTION
**Pull request by LHT129**
- refactor all simd functions, just like FP32
- add openblas for ubuntu-aarch deps

### CODE REVIEW

File: tests/test_index_old.cpp [hunk 1]
```diff
@@ -13,18 +13,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "simd/simd.h"
```
Comment (wxyucs) [inline tests/test_index_old.cpp:16]: function test cases should not depend on any internal definitions or interfaces, so don't include header files from src/

File: src/simd/sq8_simd.h [hunk 2]
```diff
@@ -72,6 +72,33 @@ SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
                      uint64_t dim);
 }  // namespace sse
 
+namespace avx {
+float
+SQ8ComputeIP(const float* query,
+             const uint8_t* codes,
+             const float* lowerBound,
```
Comment (wxyucs) [inline src/simd/sq8_simd.h]: lowerBound -> lower_bound

File: src/simd/sq8_simd.h [hunk 3]
```diff
@@ -72,6 +72,33 @@ SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
...
+             const float* diff,
+             uint64_t dim);
+float
+SQ8ComputeL2Sqr(const float* query,
+                const uint8_t* codes,
+                const float* lowerBound,
```
Comment (wxyucs) [inline src/simd/sq8_simd.h]: lowerBound -> lower_bound

File: src/simd/sq8_simd.h [hunk 4]
```diff
@@ -72,6 +72,33 @@ SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
...
+                const float* diff,
+                uint64_t dim);
+float
+SQ8ComputeCodesIP(const uint8_t* codes1,
+                  const uint8_t* codes2,
+                  const float* lowerBound,
```
Comment (wxyucs) [inline src/simd/sq8_simd.h]: lowerBound -> lower_bound

File: src/simd/sq8_simd.h [hunk 5]
```diff
@@ -72,6 +72,33 @@ SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
...
+                  const float* diff,
+                  uint64_t dim);
+float
+SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
+                     const uint8_t* codes2,
+                     const float* lowerBound,
```
Comment (wxyucs) [inline src/simd/sq8_simd.h]: lowerBound -> lower_bound

File: src/simd/avx512.cpp [hunk 6]
```diff
@@ -217,15 +96,15 @@ FP32ComputeIP(const float* query, const float* codes, uint64_t dim) {
     // process 16 floats at a time
     __m512 sum = _mm512_setzero_ps();  // initialize to 0
     for (int i = 0; i < n; ++i) {
-        __m512 a = _mm512_loadu_ps(query + i * 16);     // load 16 floats from memory
-        __m512 b = _mm512_loadu_ps(codes + i * 16);     // load 16 floats from memory
-        sum = _mm512_add_ps(sum, _mm512_mul_ps(a, b));  // accumulate the product
+        __m512 a = _mm512_loadu_ps(query + i * 16);  // load 16 floats from memory
+        __m512 b = _mm512_loadu_ps(codes + i * 16);  // load 16 floats from memory
+        sum = _mm512_fmadd_ps(a, b, sum);            // accumulate the product
```
Comment (wxyucs) [inline src/simd/avx512.cpp:101]: nice opt

File: tests/test_index_old.cpp (hunk 1 above)
Comment (LHT129) [reply tests/test_index_old.cpp:16]: some basic func like l2sqr, innerproduct is common in functest & internal, If I write it twice, is it a bit redundant? The correctness of the simd module has been guaranteed by unittest.

File: src/simd/sq8_simd.h (hunk 5 above)
Comment (LHT129) [reply src/simd/sq8_simd.h]: donw

File: src/simd/sq8_simd.h (hunk 4 above)
Comment (LHT129) [reply src/simd/sq8_simd.h]: done

File: src/simd/sq8_simd.h (hunk 3 above)
Comment (LHT129) [reply src/simd/sq8_simd.h]: done

File: src/simd/sq8_simd.h (hunk 2 above)
Comment (LHT129) [reply src/simd/sq8_simd.h]: done

File: src/simd/CMakeLists.txt [hunk 7]
```diff
@@ -22,10 +24,9 @@ if (DIST_CONTAINS_AVX)
     set (SIMD_SRCS ${SIMD_SRCS} avx.cpp)
     set_source_files_properties (avx.cpp PROPERTIES COMPILE_FLAGS "-mavx")
 endif ()
-# FIXME(LHT): cause illegal instruction on platform which has avx only
-#if (DIST_CONTAINS_AVX2)
-#    set_source_files_properties (avx.cpp PROPERTIES COMPILE_FLAGS "-mavx2 -mfma")
-#endif ()
```
Comment (inabao) [inline src/simd/CMakeLists.txt:28]: delete unuseful code

File: src/simd/CMakeLists.txt [hunk 8]
```diff
@@ -56,7 +57,7 @@ endmacro ()
 
 simd_add_definitions (DIST_CONTAINS_SSE -DENABLE_SSE=1)
 simd_add_definitions (DIST_CONTAINS_AVX -DENABLE_AVX=1)
-#simd_add_definitions (DIST_CONTAINS_AVX2 -DENABLE_AVX2=1)
```
Comment (inabao) [inline src/simd/CMakeLists.txt:59]: ditto

File: src/simd/avx2.cpp [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+
+    // Extract the result from the register
+    float result;
+    _mm_store_ss(&result, sum_final);
+
+    // Process the remaining elements
+    result += avx::SQ8ComputeL2Sqr(query + i, codes + i, lower_bound + i, diff + i, dim - i);
+    return result;
+#else
+    return vsag::avx::SQ8ComputeL2Sqr(query, codes, lower_bound, diff, dim);  // TODO
```
Comment (inabao) [inline src/simd/avx2.cpp]: `avx::SQ8ComputeL2Sqr`

File: src/simd/avx512.cpp [hunk 10]
```diff
@@ -347,28 +225,28 @@ SQ8ComputeCodesIP(const uint8_t* codes1,
...
     }
     // Horizontal addition
     float result = _mm512_reduce_add_ps(sum);
     // Process the remaining elements
-    result += avx2::SQ8ComputeCodesIP(codes1 + i, codes2 + i, lowerBound + i, diff + i, dim - i);
+    result += avx2::SQ8ComputeCodesIP(codes1 + i, codes2 + i, lower_bound + i, diff + i, dim - i);
     return result;
 #else
-    return generic::SQ8ComputeCodesIP(codes1, codes2, lowerBound, diff, dim);
+    return generic::SQ8ComputeCodesIP(codes1, codes2, lower_bound, diff, dim);
```
Comment (inabao) [inline src/simd/avx512.cpp]: why here we use `generic`

File: src/simd/avx512.cpp [hunk 11]
```diff
@@ -391,10 +269,11 @@ SQ8ComputeCodesL2Sqr(const uint8_t* codes1,
     // Horizontal addition
     float result = _mm512_reduce_add_ps(sum);
     // Process the remaining elements
-    result += avx2::SQ8ComputeCodesL2Sqr(codes1 + i, codes2 + i, lowerBound + i, diff + i, dim - i);
+    result +=
+        avx2::SQ8ComputeCodesL2Sqr(codes1 + i, codes2 + i, lower_bound + i, diff + i, dim - i);
     return result;
 #else
-    return generic::SQ8ComputeCodesL2Sqr(codes1, codes2, lowerBound, diff, dim);
+    return generic::SQ8ComputeCodesL2Sqr(codes1, codes2, lower_bound, diff, dim);
```
Comment (inabao) [inline src/simd/avx512.cpp]: ditto

File: src/simd/avx512.cpp [hunk 12]
```diff
@@ -531,6 +410,4 @@ Normalize(const float* from, float* to, uint64_t dim) {
     return norm;
 }
 
-}  // namespace avx512
-
-}  // namespace vsag
+}  // namespace vsag::avx512
```
Comment (inabao) [inline src/simd/avx512.cpp:413]: add new line?

File: src/simd/simd.cpp [hunk 13]
```diff
@@ -97,71 +59,33 @@ setup_simd() {
...
     ret.dist_support_avx512dq = true;
     ret.dist_support_avx512bw = true;
     ret.dist_support_avx512vl = true;
 #endif
-
     return ret;
 }
 
 DistanceFunc
 GetInnerProductDistanceFunc(size_t dim) {
```
Comment (inabao) [inline src/simd/simd.cpp:72]: Can we remove the functions? Just use `vsag::InnerProductDistance` replace it

File: tests/CMakeLists.txt [hunk 14]
```diff
@@ -10,9 +10,9 @@ endif ()
 if (DIST_CONTAINS_AVX)
     target_compile_definitions (unittests PRIVATE ENABLE_AVX=1)
 endif ()
-#if (DIST_CONTAINS_AVX2)
-#    target_compile_definitions (unittests PRIVATE ENABLE_AVX2=1)
-#endif ()
```
Comment (inabao) [inline tests/CMakeLists.txt:15]: ditto

File: src/simd/CMakeLists.txt (hunk 7 above)
Comment (LHT129) [inline src/simd/CMakeLists.txt:28]: what?

File: src/simd/CMakeLists.txt (hunk 8 above)
Comment (LHT129) [inline src/simd/CMakeLists.txt:59]: ditto

File: tests/CMakeLists.txt (hunk 14 above)
Comment (LHT129) [reply tests/CMakeLists.txt:15]: ditto

File: src/simd/CMakeLists.txt [hunk 15]
```diff
@@ -1,14 +1,16 @@
 set (SIMD_SRCS
         generic.cpp
+        sse.cpp
+        avx.cpp
+        avx2.cpp
+        avx512.cpp
         simd.cpp
+        basic_func.cpp
```
Comment (jiaweizone) [inline src/simd/CMakeLists.txt:8]: minor: move basic_func.cpp to L14 ?

File: src/simd/fp32_simd.cpp [hunk 16]
```diff
@@ -28,6 +28,10 @@ GetFP32ComputeIP() {
     } else if (SimdStatus::SupportAVX2()) {
 #if defined(ENABLE_AVX2)
         return avx2::FP32ComputeIP;
+#endif
+    } else if (SimdStatus::SupportAVX()) {
+#if defined(ENABLE_AVX)
```
Comment (jiaweizone) [inline src/simd/fp32_simd.cpp:33]: if not ENABLE_AVX then the if statement will be empty ?

File: src/simd/fp32_simd.cpp [hunk 17]
```diff
@@ -47,6 +51,10 @@ GetFP32ComputeL2Sqr() {
     } else if (SimdStatus::SupportAVX2()) {
 #if defined(ENABLE_AVX2)
         return avx2::FP32ComputeL2Sqr;
+#endif
+    } else if (SimdStatus::SupportAVX()) {
+#if defined(ENABLE_AVX)
+        return avx::FP32ComputeL2Sqr;
```
Comment (jiaweizone) [inline src/simd/fp32_simd.cpp:57]: ditto

File: src/simd/generic.cpp [hunk 18]
```diff
@@ -13,36 +13,29 @@
...
-
-namespace vsag {
+namespace vsag::generic {
 
 float
 L2Sqr(const void* pVect1v, const void* pVect2v, const void* qty_ptr) {
-    float* pVect1 = (float*)pVect1v;
-    float* pVect2 = (float*)pVect2v;
-    size_t qty = *((size_t*)qty_ptr);
+    auto* pVect1 = (float*)pVect1v;
```
Comment (jiaweizone) [inline src/simd/generic.cpp:22]: p_vect1 p_vect2 p_vect1v p_vect2v

File: src/simd/normalize.cpp [hunk 19]
```diff
@@ -28,6 +28,10 @@ GetNormalize() {
     } else if (SimdStatus::SupportAVX2()) {
 #if defined(ENABLE_AVX2)
         return avx2::Normalize;
+#endif
+    } else if (SimdStatus::SupportAVX()) {
+#if defined(ENABLE_AVX)
```
Comment (jiaweizone) [inline src/simd/normalize.cpp:33]: ditto

File: src/simd/sse.cpp [hunk 20]
```diff
@@ -19,281 +19,49 @@
...
-        pVect1 += 4;
-        v2 = _mm_loadu_ps(pVect2);
-        pVect2 += 4;
-        sum_prod = _mm_add_ps(sum_prod, _mm_mul_ps(v1, v2));
-    }
-    _mm_store_ps(TmpRes, sum_prod);
-    float sum = TmpRes[0] + TmpRes[1] + TmpRes[2] + TmpRes[3];
-
-    return sum;
+InnerProduct(const void* pVect1v, const void* pVect2v, const void* qty_ptr) {
```
Comment (jiaweizone) [inline src/simd/sse.cpp:38]: ditto

File: src/simd/fp32_simd.cpp (hunk 16 above)
Comment (LHT129) [reply src/simd/fp32_simd.cpp:33]: yes

File: src/simd/CMakeLists.txt (hunk 15 above)
Comment (LHT129) [reply src/simd/CMakeLists.txt:8]: I think basic_func is more importance than others, and is's used earlier

File: src/simd/avx512.cpp (hunk 12 above)
Comment (LHT129) [reply src/simd/avx512.cpp:413]: lint warning here

File: src/simd/simd.cpp (hunk 13 above)
Comment (LHT129) [reply src/simd/simd.cpp:72]: too much changes, in this pr ,only split the function, but not change usage

File: src/simd/fp32_simd.cpp (hunk 17 above)
Comment (LHT129) [reply src/simd/fp32_simd.cpp:57]: ditto

File: src/simd/generic.cpp (hunk 18 above)
Comment (LHT129) [reply src/simd/generic.cpp:22]: It's a good idea to rename all  in another PR

File: src/simd/normalize.cpp (hunk 19 above)
Comment (LHT129) [reply src/simd/normalize.cpp:33]: ditto

File: src/simd/sse.cpp (hunk 20 above)
Comment (LHT129) [reply src/simd/sse.cpp:38]: ditto

File: src/simd/avx512.cpp (hunk 11 above)
Comment (LHT129) [reply src/simd/avx512.cpp]: done

File: src/simd/avx512.cpp (hunk 10 above)
Comment (LHT129) [reply src/simd/avx512.cpp]: done

File: src/simd/avx2.cpp (hunk 9 above)
Comment (LHT129) [reply src/simd/avx2.cpp]: done

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Review by ShawnShawnYou — APPROVED** [review]

**Pull request merged**